### ISSUE TITLE
Variable Fieldpath can't reference key name with dashes

### ISSUE THREAD

**Issue by kferrone** [kubernetes-sigs/kustomize#3421]
**Describe the bug**

I have a kustomize variable which was working until a recent update. Now the variables are misbehaving for strange reasons. 

**Files that can reproduce the issue**

Here is the var in my `kustomization.yaml`
```yaml
vars:
  - name: ZONE
    objref:
      apiVersion: v1
      kind: PersistentVolume
      name: bitcoind
    fieldref:
      fieldpath: metadata.labels[failure-domain.beta.kubernetes.io/zone]
```
Then here is the corresponding PV with the label
`pv.yaml`
```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  name: bitcoind
  labels:
    failure-domain.beta.kubernetes.io/region: eu-west-2
    failure-domain.beta.kubernetes.io/zone: eu-west-2a
```
**Expected output**

I expect the variable to be able to reference any key even with special characters in the name

**Actual output**

```
Error: field specified in var '{ZONE ~G_v1_PersistentVolume {metadata.labels[failure-domain.beta.kubernetes.io/zone]}}' not found in corresponding resource
```

**Kustomize version**

```
{Version:kustomize/v3.9.1 GitCommit:7439f1809e5ccd4677ed52be7f98f2ad75122a93 BuildDate:2020-12-30T00:43:15+00:00 GoOs:darwin GoArch:amd64}
```
**Platform**

Tested on Mac and Windows, both had the issue

**Additional context**

I pretty much determined it's the dash because below works:
```yaml
vars:
  - name: ZONE
    objref:
      apiVersion: v1
      kind: PersistentVolume
      name: bitcoind
    fieldref:
      fieldpath: metadata.labels[failure_domain.beta.kubernetes.io/zone]
```
... so long as there are no dashes `-`, a key can be referenced even with dots and slashes in the name

**Comment by Shell32-Natsu** [comment]
https://github.com/kubernetes-sigs/kustomize/blob/735befef19d0128820d906bc74ae9b66aa0d2c5b/api/internal/wrappy/wnode.go#L67

In kustomize version >= v3.9, we just simply split the field path by `.`. We may improve this in future.

BTW, we are planning to remove `var` from kustomize. #2052

**Comment by diogosilva30** [comment]
how to fix this?

**Comment by aeciopires** [comment]
I had the same problem and solved it with this tip: https://github.com/kubernetes-sigs/kustomize/issues/1256#issuecomment-535007229

I used `~1` to represent a `/` 

(workaround)

**Comment by bewinter** [comment]
(Sorry for waking up this issue, but I had a hard time finding a solution for this problem as well. And since **this does not seem to be mentioned in any properly reachable/findable documentation**, I'd like to post it here for future reference.)

> I had the same problem and solved it with this tip: [#1256 (comment)](https://github.com/kubernetes-sigs/kustomize/issues/1256#issuecomment-535007229)
> 
> I used `~1` to represent a `/`
> 
> (workaround)

After a long research, much trial&error and lots of unsuccessful experiments on my side, a colleague finally told me a working solution for exactly this problem.
But, @aeciopires , sorry to inform you that this solution only works for **kustomize patches** syntax, not for **kustomize replacements**!

The solution in my case - converted to @diogosilva30's previous issue, and using some 2025 version - would look and work like this:
```yaml
        fieldPaths:
          - metadata.annotations.external-dns\.alpha\.kubernetes\.io/target
```
Hopefully, this will help future solution seekers.
